stix2/utils.py

```diff
@@ -313,18 +320,262 @@ def get_type_from_id(stix_id):
+        as a string
+    :param stix_version: A STIX version as a string
+    :return: True if the type of the given value is an SRO type of the given
+        version; False if not
+    """
+
+    result = True
+    if isinstance(value, collections.abc.Mapping):
+        value_stix_version = detect_spec_version(value)
+        if value_stix_version != stix_version:
+            result = False
+
+    if result:
+        # No need to check registration in this case
+        type_ = _stix_type_of(value)
+        result = type_ in ("sighting", "relationship")
+
+    return result
+
+
+def is_object(value, stix_version=stix2.version.DEFAULT_VERSION):
+    """
+    Determine whether an object, type, or ID is/is for any STIX object.  This
+    includes all SDOs, SCOs, meta-objects, and bundle.  If value is a type or
+    ID, this just checks whether the type was registered in the given STIX
+    version.  If a mapping, *simple* STIX version inference is additionally
+    done on the value, and the result is checked against stix_version.  It does
+    not attempt to fully validate the value.
+
+    :param value: A mapping with a "type" property, or a STIX ID or type
+        as a string
+    :param stix_version: A STIX version as a string
+    :return: True if the type of the given value is a valid STIX type with
+        respect to the given STIX version; False if not
+    """
+
+    result = True
+    if isinstance(value, collections.abc.Mapping):
+        value_stix_version = detect_spec_version(value)
+        if value_stix_version != stix_version:
+            result = False
+
+    if result:
+        cls_maps = mappings.STIX2_OBJ_MAPS[stix_version]
+        type_ = _stix_type_of(value)
+        result = type_ in cls_maps["observables"] \
+            or type_ in cls_maps["objects"]
+
+    return result
+
+
+def is_marking(value, stix_version=stix2.version.DEFAULT_VERSION):
+    """
+    Determine whether the given object, type, or ID is/is for an marking
+    definition of the given STIX version.  If value is a type or ID, this just
+    checks whether the type was registered as an SDO in the given STIX version.
+    If a mapping, *simple* STIX version inference is additionally done on the
+    value, and the result is checked against stix_version.  It does not attempt
+    to fully validate the value.
+
+    :param value: A STIX object, object ID, or type as a string.
+    :param stix_version: A STIX version as a string
+    :return: True if the value is/is for a marking definition, False otherwise.
+    """
+
+    result = True
+    if isinstance(value, collections.abc.Mapping):
+        value_stix_version = detect_spec_version(value)
+        if value_stix_version != stix_version:
+            result = False
+
+    if result:
+        # No need to check registration in this case
+        type_ = _stix_type_of(value)
+        result = type_ == "marking-definition"
+
+    return result
```

Comment:
If you pass into the function a dict and stix_version="2.1", left out the spec_version property from the dict or used the wrong value for it (assuming there are more valid values than "2.1" at some point), then it makes sense to detect the spec_version on the dict and compare it to the stix_version which was passed in, see the version discrepancy and return False.  The idea is that we want to make sure the passed in value is (a) a marking definition, and (b) conforms to the given STIX version.

It was a change I made to the original design of those functions.  I decided people would expect that those functions, with those parameters, would verify that the passed in value is of the proper type *and STIX version*.  Would you expect `is_marking(obj, stix_version="2.1")` to return `True` for a 2.0 marking definition?  I really don't think so.  I originally had written them to work only with type strings, not whole dicts (or IDs), but changing them to work with dicts really seemed to demand a bit more complexity with regards to stix version so that people don't accidentally misuse them.

The docstring definitely needs to be fixed.  Obviously there was a lot of copy-paste there, and I did have to fix that exact kind of typo in other places, but that's yet another place I missed... /sigh.

Reply:
Yes, I can see that use case, of ensuring it is both a marking definition and of the given STIX version. However, I think the two cases referenced are STIX version-agnostic. They're trying to determine if something is a marking definition of _any_ STIX version. In those cases, if a 2.0 marking definition is passed in, previously it would be detected as a marking definition. With this PR, if a 2.0 marking definition is passed in, since it's not passing a stix_version, the default of "2.1" will be used so it won't be detected as a marking definition and the outcome will be different. I'm just looking to preserve the previous behavior. To me that is preferable to making the marking functions I linked to STIX-version-aware, which would involve adding `stix_version` parameters to more functions.

Maybe if there was an option to bypass the version check in `is_marking`? Or to put it another way, to check if the given object is of the proper type and "any" STIX version.

Reply:
For `is_marking()` and `is_sro()`, when you pass a type or ID, the check actually is version-agnostic.  The spec version detection only happens when you pass a mapping.  It seemed silly to check whether "marking-definition" exists in the version-specific mapping.  We created the mappings so we know it's there, and it's not (something we intend to be) customizable so no one can change that.  It was a similar situation with SROs.  That function doesn't look up in version-specific mappings either in the type/ID case, so in that case it is also effectively spec-agnostic.

In all of other is_* functions, if you pass a type or ID, they look up the registration in the version-specific mapping.  It made sense there because they are for more complex categories of objects which vary between STIX versions, and it is possible to register custom types.

So the fact that the is_sro/marking functions behave STIX-version-agnostically in the type/ID case may be more of a coincidence and practical matter due to the way the STIX spec is.  I guess we could allow stix_version=None for all those functions and treat that as an "any spec version" case.  Then again, the same might be said of the other markings code: maybe it's a coincidence and practical matter that markings can be handled agnostically there for now, and that code doesn't need change behavior depending on STIX version.  But maybe in the future, the way they compress/expand for example, might depend on STIX version and then it won't be able to be so agnostic anymore.

Reply:
You're right, and looking closer at how those marking functions call is_marking it does look like they only pass in ids.